Use readthedocs for documentation

qiaojunfeng opened the pull request:
Close #484
Close #485

qiaojunfeng committed:
Use mike for docs versioning

- Add version warning
- Auto deploy docs to `gh-pages` on github release

The commit is kept here as a reference.

We will switch to readthedocs to deploy multiple versions of the docs,
since mike use `gh-pages` branch to host the docs, increasing the size
of the repo in the long term. See #485 for more details.

qiaojunfeng committed:
Switch to readthedocs

qiaojunfeng committed:
Update URLs

qiaojunfeng committed:
Add URL to old PDF docs & clean up github workflows

LecrisUT commented on .github/workflows/main.yml:
```diff
@@ -66,3 +66,19 @@ jobs:
+        working-directory: ./docs
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          ENABLE_MKDOCS_GIT_COMMITTERS: False
```
What is the purpose of this job? It is better to have RTD build for PR, then both author and reviewer can see the changes. If it's to test for warnings, dead links, etc. it should be specified more explicitly.

LecrisUT commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,19 @@
+build:
+  os: ubuntu-22.04
+  tools:
+    python: "3.12"
```
```suggestion
    python: "3"
```

If you don't want to manually maintain the python version and use the latest version.

LecrisUT reviewed (approved):
LGTM :+1:

Another tip, in `actions/setup-python`, `python-version: 3.x` is equivalent with the latest released version.

(the project does not appear on readthedocs.org, but I guess the backend was not setup yet)

qiaojunfeng commented:
> Another tip, in `actions/setup-python`, `python-version: 3.x` is equivalent with the latest released version.

Thanks, fixed as well.

> (the project does not appear on readthedocs.org, but I guess the backend was not setup yet)

I don't have enough permission to import the repo on RTD, already asked Giovanni to do this.

LecrisUT commented on .github/workflows/main.yml:
```diff
@@ -66,3 +66,20 @@ jobs:
+          # use the latest stable version
+          python-version: "3.x"
+          cache: "pip"
+      - run: pip install -r docs/requirements.txt
```
Oh great mkdocs is hit by the same issue that `sphinxcontrib-bibtex` has: https://github.com/mcmtroffaes/sphinxcontrib-bibtex/issues/345

```suggestion
     # `pybtex` uses `pkg_resources` which is deprecated. Use workaround until upstream `mkdocs_bibtext`decides on a solution
     # https://bitbucket.org/pybtex-devs/pybtex/issues/169/replace-pkg_resources-with
      - run: pip install -r docs/requirements.txt setuptools
```

LecrisUT replied:
Here's another issue you can follow: https://github.com/shyamd/mkdocs-bibtex/issues/228

qiaojunfeng replied:
Indeed, now I remember why I fix the python version to `3.11` in the yaml.
And thanks for creating an issue in the `mkdocs-bibtex` repo!

qiaojunfeng committed:
Fix for mkdocs-bibtex with python 3.12

Python 3.12 does not install `setuptools` in virtual env, therefore
`pkg_resources` is missing but `mkdocs-bibtex` explicitly imports it.
The solution is to explicitly install `setuptools` in the virtual env.

See https://docs.python.org/3/whatsnew/3.12.html

Co-authored-by: Cristian Le <github@lecris.me>
